File: python/mirage/core/render/json.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import json
import math
from typing import Any


def json_text(value: Any) -> str:
    """Render a value as indented JSON text.

    Args:
        value (Any): the JSON-serializable payload to render.
    """
    return json.dumps(value, ensure_ascii=False, indent=2)


def json_bytes(value: Any) -> bytes:
    """Render a value as an indented .json body.

    Every backend renders a .json leaf through here, so read() and the
    readdir-time sizing produce the same bytes for the same payload and
    the advertised size is exact by construction.

    Args:
        value (Any): the JSON-serializable payload to render.
    """
    return json_text(value).encode()


def compact_json_text(value: Any) -> str:
    """Render a value as a single line of JSON.

    Args:
        value (Any): the JSON-serializable payload to render.
    """
    return json.dumps(value, ensure_ascii=False, separators=(",", ":"))


def compact_json_bytes(value: Any) -> bytes:
    """Render a value as a single-line JSON body.

    Args:
        value (Any): the JSON-serializable payload to render.
    """
    return compact_json_text(value).encode()


def number_text(value: float) -> str:
    """Spell a float the way ECMAScript's ``Number::toString`` does.

    Python's ``repr`` and V8 agree on the shortest round-trip digits and
    differ only on layout: Python switches to an exponent below 1e-4 and
    from 1e16 and pads it to two digits (``1e-05``, ``1e+16``), where
    ECMAScript switches below 1e-6 and from 1e21 and pads nothing
    (``0.00001``, ``10000000000000000``, ``1e-7``, ``1e+21``), and an
    integral value drops its ``.0``. A non-finite value has no JSON
    spelling and renders ``null``, as ``JSON.stringify`` does. The
    typescript twin is ``String()`` itself.

    Args:
        value (float): the number to spell.
    """
    if not math.isfinite(value):
        return "null"
    if value == 0:
        return "0"
    sign = "-" if value < 0 else ""
    mantissa, _, exponent = repr(float(abs(value))).partition("e")
    whole, _, frac = mantissa.partition(".")
    digits = (whole + frac).lstrip("0")
    stripped = digits.rstrip("0")
    # value == int(stripped) * 10 ** (n - k), the spec's n and k
    k = len(stripped)
    n = len(digits) + int(exponent or 0) - len(frac)
    if k <= n <= 21:
        return sign + stripped + "0" * (n - k)
    if 0 < n <= 21:
        return f"{sign}{stripped[:n]}.{stripped[n:]}"
    if -6 < n <= 0:
        return f"{sign}0.{'0' * -n}{stripped}"
    head = stripped if k == 1 else f"{stripped[0]}.{stripped[1:]}"
    return f"{sign}{head}e{'+' if n > 1 else '-'}{abs(n - 1)}"


def _compact_value_text(value: Any) -> str:
    if isinstance(value, dict):
        items = (f"{compact_json_text(str(key))}:{_compact_value_text(item)}"
                 for key, item in value.items())
        return "{" + ",".join(items) + "}"
    if isinstance(value, list):
        return "[" + ",".join(_compact_value_text(item)
                              for item in value) + "]"
    if isinstance(value, float):
        return number_text(value)
    return compact_json_text(value)


def value_text(value: Any) -> str:
    """Spell a payload value as text, as compact JSON in both languages.

    A string is itself; anything else renders as compact JSON, so a
    boolean spells ``true`` rather than Python's ``True``, a number lays
    out as ``number_text`` says (an integral float as the integer
    TypeScript never told it apart from), and an object or array spells
    as one JSON literal rather than a ``repr``, its numbers included.
    Path labels and group values are built from this, so one collection
    grows one tree.

    Args:
        value (Any): a decoded JSON value.
    """
    if isinstance(value, str):
        return value
    return _compact_value_text(value)


def jsonl_bytes(rows: list[dict[str, Any]]) -> bytes:
    """Render rows as line-delimited JSON, one compact object per line.

    An empty row list renders as empty bytes rather than a lone newline,
    so an empty .jsonl leaf sizes and reads as a zero-byte file.

    Args:
        rows (list[dict]): the rows to render, in output order.
    """
    if not rows:
        return b""
    lines = [compact_json_text(row) for row in rows]
    return ("\n".join(lines) + "\n").encode()


def jsonl_bytes_by_created_at(rows: list[dict[str, Any]]) -> bytes:
    """Render rows as JSONL in ``created_at`` order.

    A comment feed arrives in whatever order the API paginated it, and a
    file that reads the same twice needs a stable order. ``created_at`` is
    the one field every comment normalizer emits, and a row missing it
    sorts first rather than raising.

    Args:
        rows (list[dict[str, Any]]): normalized comment rows.

    Returns:
        bytes: newline-delimited JSON, oldest first.
    """
    ordered = sorted(rows, key=lambda row: row.get("created_at") or "")
    return jsonl_bytes(ordered)

File: typescript/packages/core/src/core/render/json.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

const ENC = new TextEncoder()

export function jsonText(value: unknown): string {
  return JSON.stringify(value, null, 2)
}

// Every backend renders a .json leaf through here, so read() and the
// readdir-time sizing produce the same bytes for the same payload and the
// advertised size is exact by construction.
export function jsonBytes(value: unknown): Uint8Array {
  return ENC.encode(jsonText(value))
}

export function compactJsonText(value: unknown): string {
  return JSON.stringify(value)
}

export function compactJsonBytes(value: unknown): Uint8Array {
  return ENC.encode(compactJsonText(value))
}

/**
 * Spell a number the way ECMAScript's `Number::toString` does, which is what
 * `JSON.stringify` writes for a finite one. `String()` is the spec here; the
 * python twin reproduces its layout from `repr`, whose digits agree and whose
 * exponent thresholds and padding do not. A non-finite value has no JSON
 * spelling and renders `null`, as `JSON.stringify` does.
 */
export function numberText(value: number): string {
  return Number.isFinite(value) ? String(value) : 'null'
}

/**
 * Spell a payload value as text, as compact JSON in both languages.
 *
 * A string is itself; anything else renders as compact JSON, so a boolean
 * spells `true` rather than Python's `True`, a number lays out as
 * `numberText` says, and an object or array spells as one JSON literal rather
 * than `[object Object]`, its numbers included. Path labels and group values
 * are built from this, so one collection grows one tree.
 */
export function valueText(value: unknown): string {
  if (typeof value === 'string') return value
  if (typeof value === 'bigint') return value.toString()
  if (typeof value === 'number') return numberText(value)
  return compactJsonText(value)
}

// An empty row list renders as empty bytes rather than a lone newline, so an
// empty .jsonl leaf sizes and reads as a zero-byte file.
export function jsonlBytes(rows: readonly unknown[]): Uint8Array {
  if (rows.length === 0) return new Uint8Array()
  return ENC.encode(rows.map((row) => compactJsonText(row)).join('\n') + '\n')
}

// A comment feed arrives in whatever order the API paginated it, and a file
// that reads the same twice needs a stable order. `created_at` is the one
// field every comment normalizer emits, and a row missing it sorts first
// rather than throwing.
export function jsonlBytesByCreatedAt(rows: readonly { created_at?: string | null }[]): Uint8Array {
  const ordered = [...rows].sort((a, b) => {
    const ka = a.created_at ?? ''
    const kb = b.created_at ?? ''
    if (ka < kb) return -1
    if (ka > kb) return 1
    return 0
  })
  return jsonlBytes(ordered)
}
